### PR TITLE
Use concurrent collections instead of  synchronized methods.

### PR DESCRIPTION
**Pull request by boaks**
Use concurrent collections instead of  synchronized methods.
Add test for cancel multiple requests.

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
The downside may be, that a "request" for a Client, which is currently "deregistering", will make trouble :-).

To fix that, I would like to introduce a "pendingDeregistration" in the "org.eclipse.leshan.server.client.Client". 
This seems for me to be better than a "global synchronized". 
Any comments/opinions on that?

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java [hunk 1]
```diff
@@ -137,22 +138,23 @@ public T buildResponse(final Response coapResponse) {
...
 
-    private synchronized void addPendingRequest(String registrationId, Request coapRequest) {
+    private void addPendingRequest(String registrationId, Request coapRequest) {
         // Get or create all pending requests for this registrationId
-        ArrayList<Request> requests = pendingRequests.get(registrationId);
+        List<Request> requests = pendingRequests.get(registrationId);
         if (requests == null) {
-            requests = new ArrayList<>();
-            pendingRequests.put(registrationId, requests);
+            requests = Collections.synchronizedList(new ArrayList<Request>());
```
Comment (sbernard31) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java]: This is not thread safe.  2 threads could check that requests == null and so add the new collections twice. (Only one request will be added.

Comment (boaks) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java]: Uups sorry, I thought

            requests = Collections.synchronizedList(new ArrayList<Request>());
            List<Request> availableRequests = pendingRequests.putIfAbsent(registrationId, requests);
            if (null != availableRequests)
                requests = availableRequests;

the `putIfAbsent` would fix that.

**Comment by boaks** [comment]
From PR #100 

    Each get on ClientRegistry will return a new instance

This invalidates the proposal extending Client.

But the "race condition" is also valid for PR #100.
Any previous retrieved `Client` will cause problems when used in `send` after the deregister.
Even the synchronized `cancelPendingRequests`/`addPendingRequest` will not solve this, because you can't ensure, that someone who retrieves a `Client` finishs using it before `cancelPendingRequests` is called. And a `send` after `cancelPendingRequests` will always be handled wrong.
So we need a solution ...

**Comment by sbernard31** [comment]
I think we can close this one, right ?

**Comment by boaks** [comment]
sure.

**Pull request closed without merging**